### ISSUE TITLE
Unexpected F811 warning when defining an optional function

### ISSUE THREAD

**Issue by OddBloke** [PyCQA/pyflakes#783]
We have code that looks like this:

```py
maybe_func = None
if True:
    def maybe_func():
        return "foo"
```

and with the latest version of flake8/pyflakes, this has started emitting:

```
f811.py:3:5: F811 redefinition of unused 'maybe_func' from line 1
```

It's not entirely clear to me why: this pattern avoids an extra `else:` and indentation, and is fairly common in my experience (particularly for small functions which could be lambdas).

This also feels inconsistent, as using this pattern to assign values to variables using different syntax isn't flagged:

```py
maybe_val = None
if True:
    maybe_val = "something"
```

and you can even define your function within the `if` block, provided you call it something else and use indirect syntax to assign it to your variable:

```py
maybe_func = None
if True:
    def func_impl():
        return "foo"
    maybe_func = func_impl
```

**Comment by OddBloke** [comment]
This is also annoying from a usability perspective if you're applying flake8 to a codebase with lambdas in this pattern, e.g.:

```
maybe_func = None
if True:
    maybe_func = lambda: "foo"
```

You're first told "do not assign a lambda expression, use a def" (but, as this is assigning to `maybe_func` in an acceptable way, _not_ F811) and when you follow that advice, you're then presented with another error message in the form of F811.

**Comment by sigmavirus24** [comment]
Duplicate of https://github.com/PyCQA/pyflakes/issues/780

This is an intentional change made to improve detection in #761

**Comment by OddBloke** [comment]
The bug that #761 is fixing (#760) refers specifically to methods, not functions in general. Could the logic be modified to only apply to class bodies, rather than all function definitions?

**Comment by OddBloke** [comment]
I'd also add that unlike #780, this example does type-check successfully, but still emits an F811:

```py
import random
from typing import Optional, Callable


maybe_func: Optional[Callable[[], str]] = None
if random.choice([True, False]):
    def maybe_func() -> str:
        return "foo"

if maybe_func is not None:
    maybe_func()
```

**Comment by asottile** [comment]
#760 is also useful for functions in general -- if you move your `None` to `else:` it works well

**Comment by OddBloke** [comment]
> #760 is also useful for functions in general -- if you move your `None` to `else:` it works well

Thanks for the reply! We've made that change to unblock our CI, but it's not clear to me why it's "useful" to make that change in this context: it seems to me that it's an extra line of code which achieves the same result as our existing code.

**Comment by asottile** [comment]
redefinitions are allowed across branch arms
